### PR TITLE
refactor(api-v2): Use SparqlExtendedConstructResponse in ConstructResponseUtilV2

### LINKED ISSUE #763: refactor splitMainResourcesAndValueRdfData so it uses SparqlExtendedConstructResponse
- use `SparqlExtendedConstructResponse` so the language tag can be taken into account
- do not return a list node's label for a list value. Instead, the client has to make a request to the v2 list route (requires a change in SALSAH)
- make case class for `Map[IRI, ConstructResponseUtilV2.ResourceWithValueRdfData]`
- make case class for `Map[IRI, Seq[ValueRdfData]]`

### PR DESCRIPTION
**Pull request by benjamingeer**
This PR improves type safety in `ConstructResponseUtilV2` by using `SparqlExtendedConstructResponse`, which provides the types returned by the triplestore.

- [x] Change `SparqlExtendedConstructResponse` to use `SmartIri` for predicates.
- [x] Refactor `ConstructResponseUtilV2` and related classes.
- [x] Fix datatype typo in `v2/generateInsertStatementsForValueContent.scala.txt` (was `xsd:valueHasDecimal`, should be `xsd:decimal`).
- [x] Add upgrade script in case anyone was affected by that typo.
- [x] Add test.

Resolves #763.

### CODE REVIEW

**Comment by benjamingeer** [comment]
@tobiasschweizer Issue #763 says to:

* Take the language tag into account. But we don't use language tags in data in the triplestore, do we?
* Not return a list node's label for a list value. This was done in #1321, right?

**Comment by tobiasschweizer** [comment]
> Take the language tag into account. But we don't use language tags in data in the triplestore, do we?

No, I don't think so. We do have a language property for text values (ISO language code) but it's not the same that is used for labels etc. in the ontology.

> Not return a list node's label for a list value. This was done in #1321, right?

This was already done, yes.

**Comment by tobiasschweizer** [comment]
@benjamingeer I think I can review this PR next week. Would that be ok?

**Comment by benjamingeer** [comment]
> I think I can review this PR next week. Would that be ok?

Sure, no problem. Would you have time to look at #1259 with @andreas-aeschlimann this week?

**Comment by tobiasschweizer** [comment]
> Would you have time to look at #1259 with @andreas-aeschlimann this week?

He said he would look at it. I will remind him :-)

**Comment by benjamingeer** [comment]
> I will remind him :-)

Thanks!

**Comment by tobiasschweizer** [comment]
The update script is needed in case someone created decimal values using v2, correct?

**Comment by benjamingeer** [comment]
> The update script is needed in case someone created decimal values using v2, correct?

Yes.

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala [hunk 1]
```diff
@@ -824,20 +824,23 @@ class SearchResponderV2(responderData: ResponderData) extends ResponderWithStand
...
             // collect the IRIs of main resources returned
             mainResourceIris: Set[IRI] = searchResourceByLabelResponse.statements.foldLeft(Set.empty[IRI]) {
-                case (acc: Set[IRI], (subjIri: IRI, assertions: Seq[(IRI, String)])) =>
-                    //statement.pred == OntologyConstants.KnoraBase.IsMainResource && statement.obj.toBoolean
-
+                case (acc: Set[IRI], (subject: SubjectV2, assertions: Map[SmartIri, Seq[LiteralV2]])) =>
                     // check if the assertions represent a main resource and include its IRI if so
-                    val subjectIsMainResource: Boolean = assertions.contains((OntologyConstants.Rdf.Type, OntologyConstants.KnoraBase.Resource)) && assertions.exists {
-                        case (pred, obj) =>
-                            pred == OntologyConstants.KnoraBase.IsMainResource && obj.toBoolean
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: So with the type provided by the triplestore, no additional type conversion has to be performed, right?

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala [hunk 2]
```diff
@@ -824,20 +824,23 @@ class SearchResponderV2(responderData: ResponderData) extends ResponderWithStand
...
+                    val subjectIsMainResource: Boolean = assertions.get(OntologyConstants.KnoraBase.IsMainResource.toSmartIri) match {
+                        case Some(objs) => objs.contains(BooleanLiteralV2(true))
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: Couldn't you just check for the first element with headOption? There should be only one statement `IsMainResource`.

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 1 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: I’m on the train but if you have IntelliJ add a type annotation for `obj`, it‘s a `LiteralV2`, right? Look at the implementation of its `toBoolean` method. If the `LiteralV2` is a `BooleanLiteralV2`, the conversion has already been done in `HttpTriplestoreConnector` when the Turtle was parsed.

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 2 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: Yes, you‘re right.

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 1 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: Sorry, got confused there, ignore my reply above. But yes, the conversions are done in `HttpTriplestoreConnector`, using the types provided by the triplestore.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 3]
```diff
@@ -735,55 +940,55 @@ object ConstructResponseUtilV2 {
             case OntologyConstants.KnoraBase.IntValue =>
                 FastFuture.successful(IntegerValueContentV2(
                     ontologySchema = InternalSchema,
-                    valueHasInteger = valueObject.assertions(OntologyConstants.KnoraBase.ValueHasInteger).toInt,
+                    valueHasInteger = valueObject.requireIntObject(OntologyConstants.KnoraBase.ValueHasInteger.toSmartIri),
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: This is really a good improvement! I never liked doing these casts.

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 1 above)
Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: ok, no problem

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 2 above)
Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: ok

**Comment by benjamingeer** [comment]
> I think it would make sense to add a basic test case for ConstructResponseUtilV2. Do you think that is feasible?

I could refactor `HttpTriplestoreConnector` to make it possible to parse Turtle and construct a `SparqlExtendedConstructResponse` without actually connecting to the triplestore. Then I could add a test that took some hard-coded Turtle, parsed it, and ran it through `ConstructResponseUtilV2`. Would that be what you have in mind?

**Comment by tobiasschweizer** [comment]
> Would that be what you have in mind?

Yes, no actual connection the triplestore is needed. Also we can get an idea how much time complex results need for processing (nested resources).

File: webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala (hunk 2 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala]: Done in 3d26cf3.

**Comment by benjamingeer** [comment]
> Yes, no actual connection the triplestore is needed.

Done in 3d26cf3.

> Also we can get an idea how much time complex results need for processing (nested resources).

We can't really see this now, because `ConstructResponseUtilV2` does additional SPARQL queries, e.g. to get project info.

**Pull request merged**